From e2f829f406815188f3e11fd8801315caa33047f4 Mon Sep 17 00:00:00 2001
From: Przemyslaw Czarnota <przemyslawx.czarnota@intel.com>
Date: Wed, 4 Nov 2020 15:07:44 +0100
Subject: [PATCH] PONRTSYS-8184: extVLAN untagged drop rules not configured

- Merge only these rules which have the same act->merge_key

Investigation has shown that sometimes DROP rules are merged into PASS
rules and this is wrong. It happens because of special LCT handling,
where DROP rules are converted into PASS, and this is correct, but we
should not merge such rules into other PASS rules.

To prevent such behavior, during the merge, the merge_key is compared
and merge will not be done if merge_keys differ.
---
 drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c     | 11 ++++++++++-
 drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c |  2 +-
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
index daa6f80c2c5e..8dc82f8961e2 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
@@ -23,11 +23,17 @@
 
 static LIST_HEAD(tc_ext_vlan_storage);
 
-static bool has_cookie(struct dp_act_vlan *act)
+static bool has_cookie(const struct dp_act_vlan *act)
 {
 	return act->merge_key.hi != 0 && act->merge_key.lo != 0;
 }
 
+static bool dp_merge_key_eq(const struct dp_merge_key *a,
+			    const struct dp_merge_key *b)
+{
+	return a->hi == b->hi && a->lo == b->lo;
+}
+
 static bool is_dscp(struct dp_act_vlan *act)
 {
 	unsigned int i;
@@ -134,6 +140,9 @@ static int merge(int old_def,
 		if (!dp_pattern_vlan_eq(old_inner, new_inner))
 			return -EINVAL;
 
+	if (!dp_merge_key_eq(&old->merge_key, &new->merge_key))
+		return -EINVAL;
+
 	if (old->act != new->act)
 		return -EINVAL;
 
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
index aa030eeb5231..453d227f80d0 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
@@ -136,7 +136,7 @@ static bool fwd_drop_action_parse(struct net_device *dev,
 		if (a->ops && a->ops->type == TCA_ACT_GACT &&
 		    ((struct tcf_gact *)a)->tcf_action == TC_ACT_OK) {
 			act->act = DP_VLAN_ACT_FWD;
-			netdev_dbg(dev, "DROP action detected, act: %x\n",
+			netdev_dbg(dev, "PASS action detected, act: %x\n",
 				   act->act);
 			return true;
 		}
